flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/BarrierBuffer.java

```diff
@@ -207,11 +210,28 @@ else if (bufferOrEvent.getEvent().getClass() == CancelCheckpointMarker.class) {
 				if (bufferOrEvent.getEvent().getClass() == EndOfPartitionEvent.class) {
 					processEndOfPartition();
 				}
-				return bufferOrEvent;
+				return next;
 			}
 		}
 	}
 
+	private Optional<BufferOrEvent> handleEmptyBuffer() throws Exception {
+		if (!inputGate.isFinished()) {
+			return Optional.empty();
+		}
+
+		if (!endOfStream) {
+			// end of input stream. stream continues with the buffered data
```

Comment:
This comment and interaction is not very straight forward IMO. Could we pull out the code that continues with the buffered data into a dedicated method that is a) used as part of `pollNext()` for that case and b) is used here instead of `pollNext()`? Because we already know we will end up in a sub-case of `pollNext`.